OutOfMemory exception on opening big text files in editor

## Issue explanation (write below this line)



## Exception
* User Action: UI Error
* Request: Application crash
* Version: 3.5.3
* OS: Linux samsung/crownltexx/crownlte:10/QP1A.190711.020/N960FXXU6FTK1:user/release-keys 10 - 29
* Device: crownlte
* Model: SM-N960F
* Product: crownltexx
<details><summary><b>Crash log </b></summary><p>


java.lang.RuntimeException: An error occurred while executing doInBackground()
 at android.os.AsyncTask$4.done(AsyncTask.java:399)
 at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
 at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
 at java.util.concurrent.FutureTask.run(FutureTask.java:271)
 at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:289)
 at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
 at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
 at java.lang.Thread.run(Thread.java:919)
Caused by: java.lang.OutOfMemoryError: Failed to allocate a 174063624 byte allocation with 25149440 free bytes and 165MB until OOM, target footprint 119906520, growth limit 268435456
 at java.util.Arrays.copyOf(Arrays.java:3257)
 at java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:124)
 at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:448)
 at java.lang.StringBuilder.append(StringBuilder.java:137)
 at com.amaze.filemanager.asynchronous.asynctasks.ReadFileTask.doInBackground(ReadFileTask.java:118)
 at com.amaze.filemanager.asynchronous.asynctasks.ReadFileTask.doInBackground(ReadFileTask.java:47)
 at android.os.AsyncTask$3.call(AsyncTask.java:378)
 at java.util.concurrent.FutureTask.run(FutureTask.java:266)
 ... 4 more


</details>
<hr>

--- a/app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/ReadFileTask.java
+++ b/app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/ReadFileTask.java
@@ -51,6 +51,7 @@ public class ReadFileTask extends AsyncTask<Void, Void, ReadFileTask.ReturnedVal
   public static final int NORMAL = 0;
   public static final int EXCEPTION_STREAM_NOT_FOUND = -1;
   public static final int EXCEPTION_IO = -2;
+  public static final int EXCEPTION_OOM = -3;
 
   private ContentResolver contentResolver;
   private EditableFileAbstraction fileAbstraction;
@@ -126,6 +127,9 @@ public class ReadFileTask extends AsyncTask<Void, Void, ReadFileTask.ReturnedVal
     } catch (IOException e) {
       e.printStackTrace();
       return new ReturnedValues(EXCEPTION_IO);
+    } catch (OutOfMemoryError e) {
+      e.printStackTrace();
+      return new ReturnedValues(EXCEPTION_OOM);
     }
 
     return new ReturnedValues(stringBuilder.toString(), cachedFile);

--- a/app/src/main/java/com/amaze/filemanager/ui/activities/TextEditorActivity.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/activities/TextEditorActivity.java
@@ -49,6 +49,7 @@ import com.google.android.material.snackbar.Snackbar;
 import android.animation.Animator;
 import android.animation.AnimatorListenerAdapter;
 import android.animation.ObjectAnimator;
+import android.app.ActivityManager;
 import android.content.Context;
 import android.graphics.Color;
 import android.graphics.Typeface;
@@ -60,6 +61,7 @@ import android.text.Spanned;
 import android.text.TextWatcher;
 import android.text.style.BackgroundColorSpan;
 import android.util.DisplayMetrics;
+import android.util.Log;
 import android.view.Menu;
 import android.view.MenuItem;
 import android.view.View;
@@ -72,6 +74,8 @@ import android.widget.ImageButton;
 import android.widget.RelativeLayout;
 import android.widget.ScrollView;
 import android.widget.Toast;
+
+import androidx.documentfile.provider.DocumentFile;
 
 public class TextEditorActivity extends ThemedActivity
     implements TextWatcher, View.OnClickListener {
@@ -171,7 +175,16 @@ public class TextEditorActivity extends ThemedActivity
       mInput.setScrollY(index);
       if (savedInstanceState.getBoolean(KEY_MONOFONT)) mInput.setTypeface(mInputTypefaceMono);
     } else {
-      load();
+
+      if (isMemoryAvailable(DocumentFile.fromSingleUri(this, mFile.uri).length())) {
+        load();
+        Log.e("test", "onCreate:" + "Load");
+      } else {
+        Log.e("test", "onCreate:" + "IsLowOnMemory");
+        Toast.makeText(this, R.string.error_file_too_large, Toast.LENGTH_SHORT).show();
+        finish();
+        return;
+      }
     }
     initStatusBarResources(findViewById(R.id.texteditor));
   }
@@ -632,5 +645,23 @@ public class TextEditorActivity extends ThemedActivity
     for (BackgroundColorSpan colorSpan : colorSpans) {
       mInput.getText().removeSpan(colorSpan);
     }
+  }
+
+  private boolean isMemoryAvailable(long size) {
+    ActivityManager activityManager = (ActivityManager) this.getSystemService(ACTIVITY_SERVICE);
+    ActivityManager.MemoryInfo memoryInfo = new ActivityManager.MemoryInfo();
+    activityManager.getMemoryInfo(memoryInfo);
+
+    Log.e(
+        "test",
+        "isMemoryAvailable() called with: "
+            + "size = ["
+            + size
+            + "]\n"
+            + "availMem = ["
+            + memoryInfo.availMem
+            + "]");
+
+    return memoryInfo.availMem > size;
   }
 }

--- a/app/src/main/java/com/amaze/filemanager/ui/activities/TextEditorActivity.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/activities/TextEditorActivity.java
@@ -49,7 +49,6 @@ import com.google.android.material.snackbar.Snackbar;
 import android.animation.Animator;
 import android.animation.AnimatorListenerAdapter;
 import android.animation.ObjectAnimator;
-import android.app.ActivityManager;
 import android.content.Context;
 import android.graphics.Color;
 import android.graphics.Typeface;
@@ -61,7 +60,6 @@ import android.text.Spanned;
 import android.text.TextWatcher;
 import android.text.style.BackgroundColorSpan;
 import android.util.DisplayMetrics;
-import android.util.Log;
 import android.view.Menu;
 import android.view.MenuItem;
 import android.view.View;
@@ -74,8 +72,6 @@ import android.widget.ImageButton;
 import android.widget.RelativeLayout;
 import android.widget.ScrollView;
 import android.widget.Toast;
-
-import androidx.documentfile.provider.DocumentFile;
 
 public class TextEditorActivity extends ThemedActivity
     implements TextWatcher, View.OnClickListener {
@@ -175,16 +171,7 @@ public class TextEditorActivity extends ThemedActivity
       mInput.setScrollY(index);
       if (savedInstanceState.getBoolean(KEY_MONOFONT)) mInput.setTypeface(mInputTypefaceMono);
     } else {
-
-      if (isMemoryAvailable(DocumentFile.fromSingleUri(this, mFile.uri).length())) {
-        load();
-        Log.e("test", "onCreate:" + "Load");
-      } else {
-        Log.e("test", "onCreate:" + "IsLowOnMemory");
-        Toast.makeText(this, R.string.error_file_too_large, Toast.LENGTH_SHORT).show();
-        finish();
-        return;
-      }
+      load();
     }
     initStatusBarResources(findViewById(R.id.texteditor));
   }
@@ -330,6 +317,14 @@ public class TextEditorActivity extends ThemedActivity
                   break;
                 case ReadFileTask.EXCEPTION_IO:
                   Toast.makeText(getApplicationContext(), R.string.error_io, Toast.LENGTH_SHORT)
+                      .show();
+                  finish();
+                  break;
+                case ReadFileTask.EXCEPTION_OOM:
+                  Toast.makeText(
+                          getApplicationContext(),
+                          R.string.error_file_too_large,
+                          Toast.LENGTH_SHORT)
                       .show();
                   finish();
                   break;
@@ -645,23 +640,5 @@ public class TextEditorActivity extends ThemedActivity
     for (BackgroundColorSpan colorSpan : colorSpans) {
       mInput.getText().removeSpan(colorSpan);
     }
-  }
-
-  private boolean isMemoryAvailable(long size) {
-    ActivityManager activityManager = (ActivityManager) this.getSystemService(ACTIVITY_SERVICE);
-    ActivityManager.MemoryInfo memoryInfo = new ActivityManager.MemoryInfo();
-    activityManager.getMemoryInfo(memoryInfo);
-
-    Log.e(
-        "test",
-        "isMemoryAvailable() called with: "
-            + "size = ["
-            + size
-            + "]\n"
-            + "availMem = ["
-            + memoryInfo.availMem
-            + "]");
-
-    return memoryInfo.availMem > size;
   }
 }

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -442,6 +442,7 @@
     <string name="smb_launch_error">You will need to copy this file to storage to open it</string>
     <string name="error_file_not_found">Error: File Not Found</string>
     <string name="error_io">I/O Error</string>
+    <string name="error_file_too_large">Error: File Too Large</string>
     <string name="medium_text">Medium Text</string>
     <string name="permission_read">Read</string>
     <string name="permission_write">Write</string>

